src/checkers/inference/solver/backend/lingeling/LingelingSolver.java

```diff
@@ -33,7 +33,10 @@
 public class LingelingSolver extends MaxSatSolver {
 
     // Lingeling binary executable file should be located at JSR308/lingeling/lingeling.
-    private final String lingeling = System.getenv().get("JSR308") + "/lingeling/lingeling";
+    private String currentDirPath = System.getProperty("user.dir");
+    private String JSR308 = currentDirPath.substring(0, currentDirPath.lastIndexOf("checker-framework-inference/"));
+
+    private final String lingeling = JSR308 + "lingeling/lingeling";
```

Comment:
Lets apply 1 more best practice, use `java.nio.file.Path` and Paths class to form the full command so that the code is more robust to different operating system environments. see https://stackoverflow.com/questions/34459486/joining-paths-in-java

Reply:
OK, I will change it.

Reply:
Hello Jeff, I am not sure if we can use Path and Paths.

For the current case, we run `./gradlew testDataflowExternalSolvers` and the path of the directory that the shell script runs is `/home/user/xxxxxx/CFI/testing/dataflowexample/`. We can find that, in **LingelingSolver.java**, the path of current working directory is also the path of the directory of the shell script.

In this case, I can write these code to get the path of JSR308 and lingeling:
```
    private String currentDirPath = System.getProperty("user.dir");
    private Path JSR308 = Paths.get(currentDirPath, "../../..");
    private final String lingeling = JSR308 + "/lingeling/lingeling";
```

But, since the caller is random, we can not know what the path of the current working directory exactly is in **LingelingSolver.java** in other cases. "../../.." is a kind of hard coding and will be invalid when the working directory is changed.

Reply:
Can we assume that the environment variable **CHECKERFRAMEWORK** always exists? If so, we can use it to determine the relative path of JSR308.

Reply:
How about something like

```
private static final String CFIProjectName = "checker-framework-inference";
private String currentDirPath = System.getProperty("user.dir");
private String CFIDir = currentDirPath.substring(0, currentDirPath.lastIndexOf(CFIProjectName) + CFIProjectName.length());
private final String lingeling = Paths.get(CFIDir, "lib", "lingeling", "lingeling");
```

Reply:
Thanks Jeff, it sounds good. I will do it in this way.